### PR TITLE
Fix intersection of arrays/additional properties and time

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
Ensure properly marshalling and unmarshalling of custom time formats
when in a slice or map.

### CODE REVIEW

**Comment by jhendrixMSFT** [comment]
Unfortunately we don't have any tests in the test server that cover this case.  I had to hack an existing swagger to test the changes.  Snippets from the diff.

```
 type DatetimeWrapper struct {
+	// OPTIONAL; Contains additional key/value pairs not defined in the schema.
+	AdditionalProperties map[string]*time.Time
 	Field *time.Time `json:"field,omitempty"`
 	Now   *time.Time `json:"now,omitempty"`
+
+	// List of schedule times for backup
+	Schedules []*time.Time `json:"schedules,omitempty"`
 }
...
+	aux := make([]*timeRFC3339, len(d.Schedules), len(d.Schedules))
+	for i := 0; i < len(d.Schedules); i++ {
+		aux[i] = (*timeRFC3339)(d.Schedules[i])
+	}
+	populate(objectMap, "schedules", aux)
+	if d.AdditionalProperties != nil {
+		for key, val := range d.AdditionalProperties {
+			objectMap[key] = (*timeRFC3339)(val)
+		}
+	}
...
+		case "schedules":
+			var aux []*timeRFC3339
+			err = unpopulate(val, &aux)
+			for _, au := range aux {
+				d.Schedules = append(d.Schedules, (*time.Time)(au))
+			}
+			delete(rawMsg, key)
+		default:
+			if d.AdditionalProperties == nil {
+				d.AdditionalProperties = map[string]*time.Time{}
+			}
+			if val != nil {
+				var aux timeRFC3339
+				err = json.Unmarshal(val, &aux)
+				d.AdditionalProperties[key] = (*time.Time)(&aux)
+			}
+			delete(rawMsg, key)

**Comment by jhendrixMSFT** [comment]
Fixes https://github.com/Azure/autorest.go/issues/640

File: src/transform/transform.ts [hunk 1]
```diff
@@ -41,7 +41,7 @@ async function process(session: Session<CodeModel>) {
   // this must happen before processing objects as we depend on the
   // schema type being an actual Go type.
   for (const dictionary of values(session.model.schemas.dictionaries)) {
-    dictionary.elementType.language.go!.name = schemaTypeToGoType(session.model, dictionary.elementType, false);
+    dictionary.elementType.language.go!.name = schemaTypeToGoType(session.model, dictionary.elementType, true);
```
Comment (catalinaperalta) [inline src/transform/transform.ts:44]: Trying to understand what case we weren't covering before so that this switched to true....were we generating incorrectly for all in body params? Or are there cases where this should be false?

Comment (jhendrixMSFT) [reply src/transform/transform.ts:44]: The `inBody` param indicates if the type is passed via HTTP body or not (the "not" being a header).  When this is `true`, we add some extra annotations to the schema which is required for the intersection of additional properties and custom time times.
However, I think this parameter could possibly be removed as I don't think it matters if the extra annotations are there if we don't use them.

Comment (jhendrixMSFT) [reply src/transform/transform.ts:44]: Removing the param will erroneously add custom time types where they aren't needed, so it will remain for now.

Comment (catalinaperalta) [reply src/transform/transform.ts:44]: Ah got it, it's for additional props. Thanks!

**Review by catalinaperalta — APPROVED** [review]
LGTM. Sad there isnt a case to cover this in the testserver, maybe we should open an issue to add a test for it?

**Comment by jhendrixMSFT** [comment]
I was thinking about this.  Test holes like these are pretty unique to Go, given the way our marshallers/unmarshallers work (I believe there are other, similar corner-cases).  We could add them to the test server, however I'm wondering if it would be better if we had our own, generator-specific swagger to test all of our odd cases.  Thoughts?

**Comment by catalinaperalta** [comment]
> I was thinking about this. Test holes like these are pretty unique to Go, given the way our marshallers/unmarshallers work (I believe there are other, similar corner-cases). We could add them to the test server, however I'm wondering if it would be better if we had our own, generator-specific swagger to test all of our odd cases. Thoughts?

Yeah I agree. I think it would help us to maintain a consistent set of test cases for cases that are specifically important for Go to test. I think we've run across these a few times since developing the track2 generator, so having our own generator specific swagger would be a great idea.

**Comment by jhendrixMSFT** [comment]
Opened https://github.com/Azure/autorest.go/issues/643
